File: swift-bridge/Sources/AVAudioBridge/CompressedBuffer.swift
import AVFoundation
import Foundation

struct AudioCompressedBufferInfoPayload: Codable {
    let packetCapacity: UInt32
    let packetCount: UInt32
    let maximumPacketSize: Int
    let byteCapacity: UInt32
    let byteLength: UInt32
}

@_cdecl("av_audio_compressed_buffer_create")
public func av_audio_compressed_buffer_create(
    _ formatPtr: UnsafeMutableRawPointer,
    _ packetCapacity: UInt32,
    _ maximumPacketSize: Int,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    let format = Unmanaged<AVAudioFormat>.fromOpaque(formatPtr).takeUnretainedValue()
    let buffer = AVAudioCompressedBuffer(
        format: format,
        packetCapacity: AVAudioPacketCount(packetCapacity),
        maximumPacketSize: maximumPacketSize
    )
    return Unmanaged.passRetained(buffer).toOpaque()
}

@_cdecl("av_audio_compressed_buffer_release")
public func av_audio_compressed_buffer_release(_ ptr: UnsafeMutableRawPointer?) {
    guard let ptr else { return }
    Unmanaged<AVAudioCompressedBuffer>.fromOpaque(ptr).release()
}

@_cdecl("av_audio_compressed_buffer_info_json")
public func av_audio_compressed_buffer_info_json(
    _ ptr: UnsafeMutableRawPointer,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    let buffer = Unmanaged<AVAudioCompressedBuffer>.fromOpaque(ptr).takeUnretainedValue()
    let payload = AudioCompressedBufferInfoPayload(
        packetCapacity: buffer.packetCapacity,
        packetCount: buffer.packetCount,
        maximumPacketSize: buffer.maximumPacketSize,
        byteCapacity: buffer.byteCapacity,
        byteLength: buffer.byteLength
    )
    do {
        return ffiString(try avaEncodeJSON(payload))
    } catch {
        outError?.pointee = ffiString(error.localizedDescription)
        return nil
    }
}

@_cdecl("av_audio_compressed_buffer_set_packet_count")
public func av_audio_compressed_buffer_set_packet_count(
    _ ptr: UnsafeMutableRawPointer,
    _ packetCount: UInt32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    let buffer = Unmanaged<AVAudioCompressedBuffer>.fromOpaque(ptr).takeUnretainedValue()
    guard packetCount <= buffer.packetCapacity else {
        outError?.pointee = ffiString("packet count exceeds packet capacity")
        return AVA_INVALID_ARGUMENT
    }
    buffer.packetCount = AVAudioPacketCount(packetCount)
    return AVA_OK
}

@_cdecl("av_audio_compressed_buffer_set_byte_length")
public func av_audio_compressed_buffer_set_byte_length(
    _ ptr: UnsafeMutableRawPointer,
    _ byteLength: UInt32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    let buffer = Unmanaged<AVAudioCompressedBuffer>.fromOpaque(ptr).takeUnretainedValue()
    guard byteLength <= buffer.byteCapacity else {
        outError?.pointee = ffiString("byte length exceeds byte capacity")
        return AVA_INVALID_ARGUMENT
    }
    buffer.byteLength = byteLength
    return AVA_OK
}

@_cdecl("av_audio_compressed_buffer_copy_format")
public func av_audio_compressed_buffer_copy_format(_ ptr: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer? {
    let buffer = Unmanaged<AVAudioCompressedBuffer>.fromOpaque(ptr).takeUnretainedValue()
    return Unmanaged.passRetained(buffer.format).toOpaque()
}

File: src/compressed_buffer.rs
//! [`AVAudioCompressedBuffer`] wrappers.

#![allow(
    clippy::missing_errors_doc,
    clippy::module_name_repetitions,
    clippy::must_use_candidate
)]

use core::ffi::{c_char, c_void};
use core::ptr;

use serde::Deserialize;

use crate::buffer::AudioBufferHandle;
use crate::error::{from_swift, AVAudioError};
use crate::ffi;
use crate::format::AudioFormat;
use crate::types::AudioPacketCount;
use crate::util::parse_json_and_free;

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AudioCompressedBufferInfo {
    pub packet_capacity: AudioPacketCount,
    pub packet_count: AudioPacketCount,
    pub maximum_packet_size: usize,
    pub byte_capacity: u32,
    pub byte_length: u32,
}

/// Wraps an `AVAudioCompressedBuffer`.
pub struct AudioCompressedBuffer {
    pub(crate) ptr: *mut c_void,
}

unsafe impl Send for AudioCompressedBuffer {}

impl Drop for AudioCompressedBuffer {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { ffi::av_audio_compressed_buffer_release(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

impl AudioBufferHandle for AudioCompressedBuffer {
    fn as_buffer_ptr(&self) -> *mut c_void {
        self.ptr
    }
}

impl AudioCompressedBuffer {
    /// Allocates a compressed-audio buffer.
    pub fn new(
        format: &AudioFormat,
        packet_capacity: AudioPacketCount,
        maximum_packet_size: usize,
    ) -> Result<Self, AVAudioError> {
        let mut err: *mut c_char = ptr::null_mut();
        let ptr = unsafe {
            ffi::av_audio_compressed_buffer_create(
                format.ptr,
                packet_capacity,
                maximum_packet_size,
                &mut err,
            )
        };
        if ptr.is_null() {
            return Err(unsafe { from_swift(ffi::status::FORMAT_ERROR, err) });
        }
        Ok(Self { ptr })
    }

    /// Returns compressed-buffer metadata.
    pub fn info(&self) -> Result<AudioCompressedBufferInfo, AVAudioError> {
        let mut err: *mut c_char = ptr::null_mut();
        let json_ptr = unsafe { ffi::av_audio_compressed_buffer_info_json(self.ptr, &mut err) };
        if json_ptr.is_null() {
            return Err(unsafe { from_swift(ffi::status::OPERATION_FAILED, err) });
        }
        parse_json_and_free(json_ptr)
    }

    /// Returns the packet capacity.
    pub fn packet_capacity(&self) -> Result<AudioPacketCount, AVAudioError> {
        Ok(self.info()?.packet_capacity)
    }

    /// Returns the current packet count.
    pub fn packet_count(&self) -> Result<AudioPacketCount, AVAudioError> {
        Ok(self.info()?.packet_count)
    }

    /// Sets the current packet count.
    pub fn set_packet_count(&self, packet_count: AudioPacketCount) -> Result<(), AVAudioError> {
        let mut err: *mut c_char = ptr::null_mut();
        let status = unsafe {
            ffi::av_audio_compressed_buffer_set_packet_count(self.ptr, packet_count, &mut err)
        };
        if status != ffi::status::OK {
            return Err(unsafe { from_swift(status, err) });
        }
        Ok(())
    }

    /// Returns the maximum packet size in bytes.
    pub fn maximum_packet_size(&self) -> Result<usize, AVAudioError> {
        Ok(self.info()?.maximum_packet_size)
    }

    /// Returns the byte capacity.
    pub fn byte_capacity(&self) -> Result<u32, AVAudioError> {
        Ok(self.info()?.byte_capacity)
    }

    /// Returns the byte length.
    pub fn byte_length(&self) -> Result<u32, AVAudioError> {
        Ok(self.info()?.byte_length)
    }

    /// Sets the byte length.
    pub fn set_byte_length(&self, byte_length: u32) -> Result<(), AVAudioError> {
        let mut err: *mut c_char = ptr::null_mut();
        let status = unsafe {
            ffi::av_audio_compressed_buffer_set_byte_length(self.ptr, byte_length, &mut err)
        };
        if status != ffi::status::OK {
            return Err(unsafe { from_swift(status, err) });
        }
        Ok(())
    }

    /// Copies the compressed format metadata object.
    pub fn format(&self) -> Result<AudioFormat, AVAudioError> {
        let ptr = unsafe { ffi::av_audio_compressed_buffer_copy_format(self.ptr) };
        if ptr.is_null() {
            return Err(AVAudioError::FormatError(
                "compressed buffer did not provide a format".into(),
            ));
        }
        Ok(AudioFormat { ptr })
    }
}
